product_sequence/models/product_category.py

```diff
@@ -48,21 +48,43 @@ def _get_or_create_sequence(self, prefix):
         return sequence
 
     def write(self, vals):
+        sequence_ids_to_check = set()
         if "code_prefix" in vals:
             prefix = vals.get("code_prefix")
             if prefix:
                 for rec in self:
                     if rec.sequence_id:
-                        rec.sudo().sequence_id.prefix = prefix
+                        sequence_sudo = rec.sudo()
+                        sequence_ids_to_check.add(sequence_sudo.id)
+                        sequence_sudo.with_context(
+                            _no_sequence_prefix_check=True
+                        ).sequence_id.prefix = prefix
                     else:
                         vals["sequence_id"] = self._get_or_create_sequence(prefix).id
             else:
                 vals["sequence_id"] = False
-        return super().write(vals)
+        res = super().write(vals)
+        if sequence_ids_to_check:
+            self.env["ir.sequence"].sudo().browse(
+                sequence_ids_to_check
+            )._check_prefix_product_category()
```

Comment:
Why does the function decorated with the `api.constrains` needs to be called explicitly ?

Reply:
As we need to "sync" the prefix value on the `ir.sequence` from a change on `product.category`, the update of the sequence record before the call to super would trigger the constraint which would raise an error, as the category is not updated yet (as it will only be updated after the call to super).
Therefore, we want to "delay" the execution of the constraint check until both records are updated, and that's why we have to call it explicitly afterwards.